### PR TITLE
#912: [typescript .d.ts] Panel (closes #912)

### LINKED ISSUE #912: [typescript .d.ts] Panel
← #910

## requirements
{describe the new feature}

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by ascariandrea**
Closes #912

## Test Plan

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

- [ ] Google Chrome
- [ ] Internet Explorer

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/Panel/Panel.d.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    size?: HeaderSize,
+    content?: any,
+    title?: ReactNode | string, // TODO(typo): wtf
+    hideTitleWhenExpanded?: boolean,
+    menu?: any
+  }
+}
+
+export type PanelProps = {
+  children: ReactNode,
```
Comment (FrancescoCioria) [inline src/Panel/Panel.d.ts]: usually we use `any` for `children` but honestly I don't know exactly the reason...
@giogonzo is `ReactNode` ok?

File: src/Panel/Panel.d.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  type: 'docked-top' | 'docked-left' | 'docked-bottom' | 'docked-bottom' | 'floating',
+  header?: PanelProps.Header,
+  loading?: boolean,
+  dark?: boolean,
+  softLoading?: boolean,
+  softLoadingDelay?: boolean,
```
Comment (FrancescoCioria) [inline src/Panel/Panel.d.ts]: this should be a `number`
(to be precise, a `NonNegativeNumber` but we can't check that statically with Typescript)

File: src/Panel/Panel.d.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  className?: string,
+  clearMargin?: 'top' | 'left' | 'bottom' | 'right',
+  style?: Object
```
Comment (FrancescoCioria) [inline src/Panel/Panel.d.ts]: same: use `CSSProperties`

File: src/Panel/Panel.d.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  type Header = {
+    collapse?: {
+      direction: CollapseDirection,
+      onExpand: () => void,
+      onCollapse: () => void,
+      isCollapsed?: boolean
+    },
+    size?: HeaderSize,
+    content?: any,
+    title?: ReactNode | string, // TODO(typo): wtf
```
Comment (FrancescoCioria) [inline src/Panel/Panel.d.ts]: mmm `ReactNode` should accept `string`s, can you check this again?

Also, why isn't this `any` like `content` and `menu`?

File: src/Panel/PanelHeader.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+// TODO: generalize with Panel.Props
+export type PanelHeaderProps = {
+  collapse?: {
+    direction: PanelProps.CollapseDirection,
+    onToggleExpanded: () => void,
+    isExpanded?: boolean
+  },
+  size?: PanelProps.HeaderSize,
+  content?: any,
+  title?: ReactNode | string, // TODO(typo): wtf
```
Comment (FrancescoCioria) [inline src/Panel/PanelHeader.d.ts]: same comment as `Panel`

File: src/Panel/PanelHeader.d.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  menu?: ReactChildren
```
Comment (FrancescoCioria) [inline src/Panel/PanelHeader.d.ts]: and this one `ReactChildren`?
I think there is a lot of confusion on how to translate `t.ReactChildren` in Typescript.

@giogonzo do we have a guideline on how to handle these cases?

File: src/Panel/Panel.d.ts (hunk 4 above)
Comment (ascariandrea) [inline src/Panel/Panel.d.ts]: nope, `ReactNode` doesn't accept a `string`
```
type ReactNode = ReactChild | ReactFragment | boolean | null | undefined;
```

but I'm going to set to `any` until we'll find a right type for `children`

**Comment by ascariandrea** [comment]
Fixed!

**Pull request merged**